db/dataset.py

```diff
@@ -192,3 +193,116 @@ def delete(id):
     """Delete dataset with a specified ID."""
     with db.engine.begin() as connection:
         connection.execute("DELETE FROM dataset WHERE id = %s", (str(id),))
+
+
+def create_snapshot(dataset_id):
+    """Creates a snapshot of current version of a dataset.
+
+    Snapshots are stored as JSON and have the following structure:
+    {
+        "name": "..",
+        "description": "..",
+        "classes": [
+            {
+                "name": "..",
+                "description: "..",
+                "recordings": ["..", ...]
+            },
+            ...
+        ]
+    }
+
+    Args:
+        dataset_id (string/uuid): ID of a dataset.
+
+    Returns:
+        ID (UUID) of a snapshot that was created.
+    """
+    dataset = get(dataset_id)
+    if not dataset:
+        raise exceptions.NoDataFoundException("Can't find dataset with a specified ID.")
+    snapshot = {
+        "name": dataset["name"],
+        "description": dataset["description"],
+        "classes": dataset["classes"],
+    }
+    with db.engine.connect() as connection:
+        result = connection.execute(sqlalchemy.text("""
+            INSERT INTO dataset_snapshot (id, dataset_id, data)
+                 VALUES (uuid_generate_v4(), :dataset_id, :data)
+              RETURNING id
+        """), {
+            "dataset_id": dataset_id,
+            "data": json.dumps(snapshot),
+        })
+        return result.fetchone()["id"]
+
+
+def get_snapshot(id):
+    """Get snapshot of a dataset.
+
+    Args:
+        id (string/uuid): ID of a snapshot.
+
+    Returns:
+        dictionary: {
+            "id": <ID of the snapshot>,
+            "dataset_id": <ID of the dataset that this snapshot is associated with>,
+            "created": <creation time>,
+            "data": <actual content of a snapshot (see `create_snapshot` function)>
+        }
+    """
+    with db.engine.connect() as connection:
+        result = connection.execute(sqlalchemy.text("""
+            SELECT id::text
+                 , dataset_id::text
+                 , data
+                 , created
+              FROM dataset_snapshot
+             WHERE id = :id
+        """), {"id": id})
+        row = result.fetchone()
+        if not row:
+            raise db.exceptions.NoDataFoundException("Can't find dataset snapshot with a specified ID.")
```

Comment:
I see you have a `TODO` also to decide if we raise an exception on dataset `get`. My feeling for this is that we shouldn't have exceptions - they're for _exceptional_ circumstances, and data not existing isn't really exceptional in my view.
In other places in AB we don't raise exceptions here.
From my experience in other places (e.g. pymusicbrainzngs) raising exceptions is just a pain. None or an empty list tends to work better.